actionpack/test/controller/request_forgery_protection_test.rb

```diff
@@ -506,6 +506,21 @@ def test_should_block_post_with_origin_checking_and_wrong_origin
     ActionController::Base.logger = old_logger
   end
 
+  def test_should_raise_for_an_origin_header_different_of_request_base_url
+    forgery_protection_origin_check do
+      session[:_csrf_token] = @token
+      @controller.stub :form_authenticity_token, @token do
+        @request.set_header "HTTP_ORIGIN", "http://bad.host"
+        post :index, params: { custom_authenticity_token: @token }
+      rescue ActionController::InvalidAuthenticityToken => exception
+        assert_match(
+          "HTTP Origin header (http://bad.host) didn't match request.base_url (http://test.host)",
+          exception.message
+        )
+      end
```

Comment:
I wanted to use the syntax `assert_raise` but with this code:
```ruby
  def test_should_raise_for_an_origin_header_different_of_request_base_url
    forgery_protection_origin_check do
      session[:_csrf_token] = @token
      @controller.stub :form_authenticity_token, @token do
        exception = assert_raises(ActionController::InvalidAuthenticityToken) do
          @request.set_header "HTTP_ORIGIN", "http://bad.host"
          post :index, params: { custom_authenticity_token: @token }
        end
        assert_match(
          "HTTP Origin header (http://bad.host) didn't match request.base_url (http://test.host)",
          exception.message
        )
      end
    end
  end
```

I still had one failure:
```
Failure:
RequestForgeryProtectionControllerUsingResetSessionTest#test_should_raise_for_an_origin_header_different_of_request_base_url [/Users/benoit/code/rails/actionpack/test/controller/request_forgery_protection_test.rb:513]:
ActionController::InvalidAuthenticityToken expected but nothing was raised.
```

Reply:
Are you sure the Exception is raised now 😄 ?

Reply:
Ah, it doesn't raise, because it's blocked (whatever that means), like in the test above: `test_should_block_post_with_origin_checking_and_wrong_origin`

Reply:
Hmm, `assert_blocked` asserts the error is raised 🤷 
```
  def assert_blocked
    assert_raises(ActionController::InvalidAuthenticityToken) do
      yield
    end
end
```

Reply:
Had another look 😄 

This test is in `RequestForgeryProtectionTests` that is included in multiple test classes, for example:
```ruby
class RequestForgeryProtectionControllerUsingExceptionTest < ActionController::TestCase
  include RequestForgeryProtectionTests
```

But as each strategy has a different behaviour some methods are overridden.
For example `ActionController::InvalidAuthenticityToken` isn't raised for `null_session`. 
So `assert_blocked` is overridden for example.

It think this test should be moved to the ﻿﻿RequestForgeryProtectionControllerUsingExceptionTest as that is the only one that raises exceptions in this case:
https://github.com/rails/rails/blob/main/actionpack/test/controller/request_forgery_protection_test.rb#L702-L709

There we can just do `assert_raises`, you tried earlier on.

Reply:
That's a great suggestion. It is much much better that way. I have implemented your proposal.